widget/widget.go

```diff
@@ -13,51 +15,95 @@ type BaseWidget struct {
 	position fyne.Position
 	Hidden   bool
 
-	impl fyne.Widget
+	impl         fyne.Widget
+	propertyLock sync.RWMutex
 }
 
 // ExtendBaseWidget is used by an extending widget to make use of BaseWidget functionality.
 func (w *BaseWidget) ExtendBaseWidget(wid fyne.Widget) {
+	w.propertyLock.Lock()
+	defer w.propertyLock.Unlock()
+
 	if w.impl != nil {
 		return
 	}
 
 	w.impl = wid
 }
 
+// ReadFields provides a guaranteed thread safe way to access widget fields.
+func (w *BaseWidget) ReadFields(f func()) {
+	w.propertyLock.RLock()
+	defer w.propertyLock.RUnlock()
+
+	f()
+}
+
+// SetFieldsAndRefresh helps to make changes to a widget that should be followed by a refresh.
+// This method is a guaranteed thread-safe way of directly manipulating widget fields.
+func (w *BaseWidget) SetFieldsAndRefresh(f func()) {
+	w.propertyLock.Lock()
+	f()
+	w.propertyLock.Unlock()
+
+	w.Refresh()
+}
+
 // Size gets the current size of this widget.
 func (w *BaseWidget) Size() fyne.Size {
+	w.propertyLock.RLock()
+	defer w.propertyLock.RUnlock()
```

Comment:
Use ReadFields(f func(){...}) for consistency

Reply:
Yes, the return here meant that it would double the number of variables.
I'm thinking about a ReadAndReturn(func()) interface{} helper for these situations

Reply:
Not sure about that - the type cast may be uglier than extra variables

Reply:
Possible. I was going to play with it and see